Keycloak integration is broken

When completing the demo, redirection to the app fails with this error in Keycloak:
```
vc-authn-keycloak-1       | 2023-09-13 19:59:10,314 ERROR [org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider] (executor-thread-3) Failed to make identity provider oauth callback: org.keycloak.broker.provider.IdentityBrokerException: Could not decode access token response.
vc-authn-keycloak-1       |     at org.keycloak.broker.oidc.OIDCIdentityProvider.getFederatedIdentity(OIDCIdentityProvider.java:370)
vc-authn-keycloak-1       |     at org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider$Endpoint.authResponse(AbstractOAuth2IdentityProvider.java:505)
vc-authn-keycloak-1       |     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
vc-authn-keycloak-1       |     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
vc-authn-keycloak-1       |     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
vc-authn-keycloak-1       |     at java.base/java.lang.reflect.Method.invoke(Method.java:568)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:154)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:118)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:560)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:452)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:413)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:415)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:378)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:174)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:142)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:168)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:131)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:33)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:429)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:240)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:154)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:157)
vc-authn-keycloak-1       |     at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:229)
vc-authn-keycloak-1       |     at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:82)
vc-authn-keycloak-1       |     at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:147)
vc-authn-keycloak-1       |     at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:84)
vc-authn-keycloak-1       |     at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:44)
vc-authn-keycloak-1       |     at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
vc-authn-keycloak-1       |     at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
vc-authn-keycloak-1       |     at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
vc-authn-keycloak-1       |     at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1.handle(HttpServerCommonHandlers.java:58)
vc-authn-keycloak-1       |     at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1.handle(HttpServerCommonHandlers.java:36)
vc-authn-keycloak-1       |     at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
vc-authn-keycloak-1       |     at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
vc-authn-keycloak-1       |     at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
vc-authn-keycloak-1       |     at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
vc-authn-keycloak-1       |     at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
vc-authn-keycloak-1       |     at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
vc-authn-keycloak-1       |     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
vc-authn-keycloak-1       |     at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
vc-authn-keycloak-1       |     at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
vc-authn-keycloak-1       |     at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
vc-authn-keycloak-1       |     at java.base/java.lang.Thread.run(Thread.java:833)
vc-authn-keycloak-1       | Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Internal': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
vc-authn-keycloak-1       |  at [Source: (String)"Internal Server Error"; line: 1, column: 9]
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2477)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:760)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:3041)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:3019)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._matchToken(ReaderBasedJsonParser.java:2793)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:2066)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:808)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.databind.ObjectMapper._initForReading(ObjectMapper.java:4912)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4818)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3772)
vc-authn-keycloak-1       |     at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3740)
vc-authn-keycloak-1       |     at org.keycloak.util.JsonSerialization.readValue(JsonSerialization.java:75)
vc-authn-keycloak-1       |     at org.keycloak.broker.oidc.OIDCIdentityProvider.getFederatedIdentity(OIDCIdentityProvider.java:368)
vc-authn-keycloak-1       |     ... 44 more
vc-authn-keycloak-1       | 
vc-authn-keycloak-1       | 
vc-authn-keycloak-1       | 2023-09-13 19:59:10,316 WARN  [org.keycloak.events] (executor-thread-3) type=IDENTITY_PROVIDER_LOGIN_ERROR, realmId=vc-authn, clientId=vue-fe, userId=null, ipAddress=172.23.0.1, error=identity_provider_login_failure, code_id=4c0d6cbf-0b62-40c3-adcc-a060ed2ecaaa
```

--- a/docker/docker-compose.yaml
+++ b/docker/docker-compose.yaml
@@ -68,7 +68,7 @@ services:
       #
       # The following parameter addresses a redirect error on logout in later (post KC 16 at least).
       # It may be fixable by updating the vue app as well -- google "keycloak error Invalid parameter: redirect_uri"
-      KC_SPI_LOGIN_PROTOCOL_OPENID_CONNECT_LEGACY_LOGOUT_REDIRECT_URI: true
+      KC_SPI_LOGIN_PROTOCOL_OPENID_CONNECT_LEGACY_LOGOUT_REDIRECT_URI: "true"
     depends_on:
       - keycloak-db
     networks:
@@ -98,7 +98,7 @@ services:
       - ACAPY_WALLET_STORAGE_TYPE=${WALLET_TYPE}
       - ACAPY_READ_ONLY_LEDGER=true
       - ACAPY_GENESIS_TRANSACTIONS_LIST=/tmp/ledgers.yaml
-      - ACAPY_LOG_LEVEL=debug
+      - ACAPY_LOG_LEVEL=info
       - ACAPY_WEBHOOK_URL=${CONTROLLER_WEB_HOOK_URL}
       - ACAPY_AUTO_PROVISION=true
       - POSTGRESQL_WALLET_HOST=${POSTGRESQL_WALLET_HOST}

--- a/oidc-controller/api/core/acapy/client.py
+++ b/oidc-controller/api/core/acapy/client.py
@@ -1,11 +1,13 @@
-import requests
 import json
-import structlog
 from typing import Optional, Union
 from uuid import UUID
-from .models import WalletDid, CreatePresentationResponse
+
+import requests
+import structlog
+
 from ..config import settings
 from .config import AgentConfig, MultiTenantAcapy, SingleTenantAcapy
+from .models import CreatePresentationResponse, WalletDid
 
 _client = None
 logger = structlog.getLogger(__name__)
@@ -67,7 +69,7 @@ class AcapyClient:
         assert resp_raw.status_code == 200, resp_raw.content
         resp = json.loads(resp_raw.content)
 
-        logger.debug("<<< get_presentation_request -> {resp}")
+        logger.debug(f"<<< get_presentation_request -> {resp}")
         return resp
 
     def verify_presentation(self, presentation_exchange_id: Union[UUID, str]):

--- a/oidc-controller/api/core/models.py
+++ b/oidc-controller/api/core/models.py
@@ -45,3 +45,8 @@ class TimestampModel(BaseModel):
 
 class GenericErrorMessage(BaseModel):
     detail: str
+
+
+class RevealedAttribute(BaseModel):
+    sub_proof_index: int
+    values: dict

--- a/oidc-controller/api/core/oidc/issue_token_service.py
+++ b/oidc-controller/api/core/oidc/issue_token_service.py
@@ -57,24 +57,34 @@ class Token(BaseModel):
 
         referent: str
         requested_attr: ReqAttr
-        for referent, requested_attr in auth_session.presentation_exchange[
-            "presentation_request"
-        ]["requested_attributes"].items():
-            logger.debug(f"Referent: {referent}, ReqAttr: {requested_attr}")
-            revealed_attrs: Dict[
-                str, RevealedAttribute
-            ] = auth_session.presentation_exchange["presentation"]["requested_proof"][
-                "revealed_attr_groups"
-            ]
-            logger.debug(f"revealed Attrs: {revealed_attrs}")
-            # loop through each value and put it in token as a claim
-            for attr_name in requested_attr["names"]:
-                logger.debug(f"AttrName: {attr_name}")
-                presentation_claims[attr_name] = Claim(
-                    type=attr_name,
-                    value=revealed_attrs[referent]["values"][attr_name]["raw"],
+        try:
+            for referent, requested_attr in auth_session.presentation_exchange[
+                "presentation_request"
+            ]["requested_attributes"].items():
+                logger.debug(
+                    f"Processing referent: {referent}, requested_attr: {requested_attr}"
                 )
-                logger.debug(f"Claims: {presentation_claims}")
+                revealed_attrs: Dict[
+                    str, RevealedAttribute
+                ] = auth_session.presentation_exchange["presentation"][
+                    "requested_proof"
+                ][
+                    "revealed_attr_groups"
+                ]
+                logger.debug(f"revealed_attrs: {revealed_attrs}")
+                # loop through each value and put it in token as a claim
+                for attr_name in requested_attr["names"]:
+                    logger.debug(f"AttrName: {attr_name}")
+                    presentation_claims[attr_name] = Claim(
+                        type=attr_name,
+                        value=revealed_attrs[referent]["values"][attr_name]["raw"],
+                    )
+                    logger.debug(f"Compiled presentation_claims: {presentation_claims}")
+        except Exception as err:
+            logger.error(
+                f"An exception occurred while extracting the proof claims: {err}"
+            )
+            raise RuntimeError(err)
 
         # look at all presentation_claims and one should
         #   match the configured subject_identifier

--- a/oidc-controller/api/core/oidc/issue_token_service.py
+++ b/oidc-controller/api/core/oidc/issue_token_service.py
@@ -149,5 +149,3 @@ class Token(BaseModel):
             result[key] = value
 
         return result
-
-        return result

--- a/oidc-controller/api/core/oidc/issue_token_service.py
+++ b/oidc-controller/api/core/oidc/issue_token_service.py
@@ -1,13 +1,16 @@
-import structlog
+import dataclasses
 import json
 import uuid
-import dataclasses
 from datetime import datetime
-from typing import List, Dict, Any
-from pydantic import BaseModel
+from typing import Any, Dict, List
+
+import structlog
 from oic.oic.message import OpenIDSchema
+from pydantic import BaseModel
+
 from ...authSessions.models import AuthSession
-from ...verificationConfigs.models import VerificationConfig
+from ...verificationConfigs.models import ReqAttr, VerificationConfig
+from ..models import RevealedAttribute
 
 logger = structlog.getLogger(__name__)
 
@@ -52,16 +55,26 @@ class Token(BaseModel):
             ]
         )
 
+        referent: str
+        requested_attr: ReqAttr
         for referent, requested_attr in auth_session.presentation_exchange[
             "presentation_request"
         ]["requested_attributes"].items():
+            logger.debug(f"Referent: {referent}, ReqAttr: {requested_attr}")
+            revealed_attrs: Dict[
+                str, RevealedAttribute
+            ] = auth_session.presentation_exchange["presentation"]["requested_proof"][
+                "revealed_attr_groups"
+            ]
+            logger.debug(f"revealed Attrs: {revealed_attrs}")
             # loop through each value and put it in token as a claim
-            revealed_attrs: Dict[str, Any] = auth_session.presentation_exchange[
-                "presentation"
-            ]["requested_proof"]["revealed_attrs"]
-            presentation_claims[requested_attr["name"]] = Claim(
-                type=requested_attr["name"], value=revealed_attrs[referent]["raw"]
-            )
+            for attr_name in requested_attr["names"]:
+                logger.debug(f"AttrName: {attr_name}")
+                presentation_claims[attr_name] = Claim(
+                    type=attr_name,
+                    value=revealed_attrs[referent]["values"][attr_name]["raw"],
+                )
+                logger.debug(f"Claims: {presentation_claims}")
 
         # look at all presentation_claims and one should
         #   match the configured subject_identifier
@@ -124,5 +137,7 @@ class Token(BaseModel):
         # add to the top level of the dict.
         for key, value in standard_claims.to_dict().items():
             result[key] = value
+
+        return result
 
         return result

--- a/oidc-controller/api/routers/oidc.py
+++ b/oidc-controller/api/routers/oidc.py
@@ -171,4 +171,5 @@ async def post_token(request: Request, db: Database = Depends(get_db)):
     token_response = provider.provider.handle_token_request(
         data, request.headers, claims
     )
+    logger.debug(f"Token response: {token_response.to_dict()}")
     return token_response.to_dict()

